pkg/api/handlers/oauth/token.go

```diff
@@ -0,0 +1,216 @@
+		return types.NewErrBadRequest("%v", Error{
+			Code:        ErrInvalidRequest,
+			Description: fmt.Sprintf("grant_type must be one of %s, not %s", strings.Join(h.oauthConfig.GrantTypesSupported, ", "), grantType),
+		})
+	}
+
+	if !slices.Contains(client.Spec.Manifest.GrantTypes, grantType) {
+		return types.NewErrBadRequest("%v", Error{
+			Code:        ErrInvalidRequest,
+			Description: "client is not allowed to use authorization_code grant type",
+		})
+	}
+
+	if grantType == "authorization_code" {
+		return h.doAuthorizationCode(req, client, req.FormValue("code"), req.FormValue("code_verifier"))
+	}
+
+	return h.doRefreshToken(req, client, req.FormValue("refresh_token"))
+}
+
+func (h *handler) doAuthorizationCode(req api.Context, oauthClient v1.OAuthClient, code, codeVerifier string) error {
+	if code == "" {
+		return types.NewErrBadRequest("%v", Error{
+			Code:        ErrInvalidRequest,
+			Description: "code is required",
+		})
+	}
+
+	var oauthAuthRequestList v1.OAuthAuthRequestList
+	if err := req.Storage.List(req.Context(), &oauthAuthRequestList, &kclient.ListOptions{
+		FieldSelector: fields.SelectorFromSet(selectors.RemoveEmpty(map[string]string{
+			"spec.hashedAuthCode": fmt.Sprintf("%x", sha256.Sum256([]byte(code))),
+		})),
+	}); err != nil {
+		return err
+	}
+	if len(oauthAuthRequestList.Items) != 1 {
+		return types.NewErrBadRequest("%v", Error{
+			Code:        ErrInvalidRequest,
+			Description: "code is invalid",
+		})
+	}
+
+	oauthAuthRequest := oauthAuthRequestList.Items[0]
+
+	// Authorization codes are one-time use
+	if err := req.Storage.Delete(req.Context(), &oauthAuthRequest); err != nil {
+		// Don't return an error if we can't delete the auth request
+		log.Warnf("failed to delete auth request: %v", err)
+	}
+
+	if oauthAuthRequest.Spec.CodeChallenge != "" {
+		switch oauthAuthRequest.Spec.CodeChallengeMethod {
+		case "S256":
+			hashedCodeVerifier := sha256.Sum256([]byte(codeVerifier))
+			if oauthAuthRequest.Spec.CodeChallenge != base64.RawURLEncoding.EncodeToString(hashedCodeVerifier[:]) {
+				return types.NewErrBadRequest("%v", Error{
+					Code:        ErrInvalidRequest,
+					Description: "code_verifier is invalid",
+				})
+			}
+		case "plain":
+			if oauthAuthRequest.Spec.CodeChallenge != codeVerifier {
+				return types.NewErrBadRequest("%v", Error{
+					Code:        ErrInvalidRequest,
+					Description: "code_verifier is invalid",
+				})
+			}
+		default:
+			return types.NewErrBadRequest("%v", Error{
+				Code:        ErrInvalidRequest,
+				Description: "code_challenge_method must be S256 or plain. ",
+			})
+		}
+	}
+
+	tkn, accessToken, err := h.gatewayClient.NewAuthTokenWithExpiration(req.Context(), oauthAuthRequest.Spec.AuthProviderNamespace, oauthAuthRequest.Spec.AuthProviderName, oauthAuthRequest.Spec.UserID, time.Hour)
+	if err != nil {
+		return fmt.Errorf("failed to create auth token: %w", err)
+	}
+
+	refreshToken := strings.ToLower(rand.Text() + rand.Text())
```

Comment:
curious is there a reason it has to be two rand.Text() appended?

Reply:
Just for more randomness.